### PR TITLE
Update OCs after ER Permissions are Revoked

### LINKED ISSUE #1157: Profile badges should be removed if E2Es have been removed
image

### PR DESCRIPTION
**Pull request by macanudo527**
#### OCs still have products from exchanges even after ER permission is revoked

- Closes #1157 

If ER permissions are revoked, users are still able to shop for variants at a distributor.

#### OC shouldn't have variant or exchange from the parent of the ER that has been revoked.

To reproduce this error, create an OC with an enterprise and add a variant. Remove the ER to add_to_order_cycle. The variant count remains the same and users can still shop for the variant on the frontend.

#### Release notes

Changelog Category: User facing changes

Update OCs after ER Permissions are Revoked

The title of the pull request will be included in the release notes.

#### Dependencies

None

#### Documentation updates

None

### CODE REVIEW

**Comment by macanudo527** [comment]
Currently, this PR deletes the exchanges and exchange_variants, but then creates another exchange_variant and touches the `enterprise` twice. It needs improvements before merge, but I can't find the problem. I'm seeking feedback on it.

**Comment by jibees** [comment]
This one is not ready for review, I'll move it in `In Dev` state.

@macanudo527 If you feel like you're blocked, or if you have any questions, feel free to start a discussion on our [Slack](https://openfoodnetwork.slack.com/archives/C2GQ45KNU)

**Comment by macanudo527** [comment]
This should be ready now. I apologize for the sloppy commits. Let me know if anything needs changed.

File: spec/models/enterprise_relationship_spec.rb
Comment (mkllnk) [inline spec/models/enterprise_relationship_spec.rb:258]:
```suggestion

    describe "updating order cycles" do
```

File: spec/models/enterprise_relationship_spec.rb
Comment (mkllnk) [inline spec/models/enterprise_relationship_spec.rb:311]: I think this would be clearer:

```suggestion

          before { er.permissions_list -= [:add_to_order_cycles]; er.save! }

```

File: app/models/enterprise_relationship.rb
Comment (mkllnk) [inline app/models/enterprise_relationship.rb:115]: I wonder if this would be better in EnterpriseRelationshipPermission.

File: app/services/exchange_deleter.rb
Comment (mkllnk) [inline app/services/exchange_deleter.rb:5]: So if I change the date of an inactive order cycle then I could make it active and still have those products in there? :thinking:

File: app/services/exchange_deleter.rb
Comment (mkllnk) [inline app/services/exchange_deleter.rb:16]: These are the outgoing variants, right? Maybe if an incoming Exchange is destroyed, it should destroy any related outgoing Exchange. And each Exchange can destroy its related ExchangeVariants. That spreads the code a bit but sounds more robust to me. And each part would be much easier to understand.

**Review by mkllnk — COMMENTED** [review]
Great, thank you. I like your test-driven approach. You are on the right track here but I wonder if the solution can be simplified by moving the responsibility to destroy related records into each model. What do you think?

File: spec/models/enterprise_relationship_spec.rb
Comment (macanudo527) [reply spec/models/enterprise_relationship_spec.rb]: Can I write a function that takes a custom operator? I tried to write something like [what is written for the assignment operator](https://github.com/openfoodfoundation/openfoodnetwork/blob/ad9e82e97364de2d4f088834eb2e954b5d072b7e/app/models/enterprise_relationship.rb#L74), but no luck. What do you think of the following?
```
er.revoke_permission!(:add_to_order_cycle)
```
and define the function in `app/models/enterprise_relationship.rb`

File: app/models/enterprise_relationship.rb
Comment (macanudo527) [reply app/models/enterprise_relationship.rb:115]: That could work. We need `destroy_related_exchanges` in `EnterpriseRelationship` to delete them if an `exchange` is destroyed, though. We would need to have similar code in two places, but it would be clearer.

File: app/services/exchange_deleter.rb
Comment (macanudo527) [reply app/services/exchange_deleter.rb:5]: Good point. I could add code that would delete the exchanges in that case, but we wouldn't be able to see the full order cycle history.

Or just destroy all exchanges even if they are expired. However, in that case we won't be able to access the history of the order cycle.

I don't know enough about the use cases to know which would be the better solution.

File: spec/models/enterprise_relationship_spec.rb
Comment (mkllnk) [reply spec/models/enterprise_relationship_spec.rb]: Perfect!

File: app/services/exchange_deleter.rb
Comment (mkllnk) [reply app/services/exchange_deleter.rb:5]: Hm, we have that history problem already. For your PR, we should remove the items from all order cycles.

One common workflow is to clone a closed order cycle and set new dates to start a new one. Then it would make sense if any revoked products were already removed from the old order cycle.

If we need the feature of knowing which products were offered through an order cycle in the past then we need to implement that separately.

File: app/services/exchange_deleter.rb
Comment (macanudo527) [reply app/services/exchange_deleter.rb]: I don't think this is possible. When you create an order cycle it creates an incoming exchange for each enterprise it is distributing for. Each of those has its own variants of course. It then creates an 'internal' outgoing exchange that combines all of those variants. This internal exchange is used for processing fees (I think). If you destroy the 'internal' outgoing exchange you would essentially destroy the whole order cycle since it would destroy all variants even ones not connected to the incoming exchange that was destroyed.

You have to destroy only the variants linked to the incoming exchange and keep the 'internal' outgoing exchange intact.

Comment (macanudo527) [reply app/services/exchange_deleter.rb]: Now that I rethink this. In the exchange model, I can just destroy the variants from the internal exchange when an incoming exchange is destroyed, leaving the internal exchange itself in place.  Maybe that is what you meant.

Comment (macanudo527) [reply app/services/exchange_deleter.rb]: Actually I ended up pushing it down into the `ExchangeVariant` model. Let me know if I should add anymore tests.

Comment (mkllnk) [reply app/services/exchange_deleter.rb]: Thank you for the detailed explanation. Even after many years of working with OFN code I still remember only my most recent work and forget the details of other parts of the code base. Your explanation sounds solid and I'll look at your new commits.

File: app/services/variant_deleter.rb
Comment (mkllnk) [inline app/services/variant_deleter.rb:21]: I shouldn't look at this at the end of my work day. What are the parent variants variants here? Aren't you selecting ExchangeVariants by variant_id and then plucking that variant_id into selecting ExchangeVariants again? Would it be the same as this?
```suggestion
    internal_variants = ExchangeVariant.where(variant_id: variant_id)
      joins(:exchange).
      where(
        exchanges: { sender: receiver, receiver: receiver, incoming: false }
      )
```

File: app/services/exchange_deleter.rb
Comment (mkllnk) [inline app/services/exchange_deleter.rb:8]: Do you still need the variable `main_exchanges`? Maybe you could call the method `delete_incoming`. And would it be useful to define `incoming: true` in the query? I can't think of a scenario where an outgoing or internal exchange would have these parent and receiver values but it could be another safeguard and even speed up the query with an indexed boolean variable.

**Review by mkllnk — COMMENTED** [review]
I'm still not sure if I understand it all but let me know if my comments make sense.

**Comment by macanudo527** [comment]
Ok, I incorporated your recent suggestions and the tests pass. I can sometimes get too hasty and not refactor. Thanks for the feedback.

File: app/models/enterprise_relationship.rb
Comment (mkllnk) [inline app/models/enterprise_relationship.rb:89]: While I thought that this method is a good addition, it's only used once in a spec. If we don't have a use for it in production code then we shouldn't have this method, I'm realising now.

File: app/services/exchange_deleter.rb
Comment (mkllnk) [inline app/services/exchange_deleter.rb:8]: This is a very small service. It could probably just be on the Permission model.

File: app/services/exchange_deleter.rb
Comment (mkllnk) [inline app/services/exchange_deleter.rb:6]: Do we need this join now?

File: app/models/exchange_variant.rb
Comment (mkllnk) [inline app/models/exchange_variant.rb:9]: I think that instead of the receiver, we should go by the order cycle. In theory, this should work fine but I keep wondering if there may be a case in the future where we delete an incoming exchange to one order cycle and not want to touch the exchange variants of other order cycles. So this ExchangeVariant has an OrderCycle through the Exchange. And we want to delete all ExchangeVariants of this Spree::Variant within this order cycle, if the destroyed exchange was an incoming one. Does that make sense?

**Review by mkllnk — CHANGES_REQUESTED** [review]
I think that my head was a bit clearer reviewing today. There's still some cleanup to do and let me know if you think that my logic to tie to an order cycle makes sense. There have been so many changes that I would be happy for you to rewrite the history with Git rebase and remove all the detours.

File: app/models/exchange_variant.rb
Comment (macanudo527) [reply app/models/exchange_variant.rb:9]: > In theory, this should work fine but I keep wondering if there may be a case in the future where we delete an incoming exchange to one order cycle and not want to touch the exchange variants of other order cycles.

I got a little confused here. 'this' is referring to the current implementation and not your suggested implementation of using the order cycle to destroy variants, right?

Comment (mkllnk) [reply app/models/exchange_variant.rb]: Yes, that's right.

**Comment by macanudo527** [comment]
I think everything has been resolved with this. Do let me know if anything looks out of place. 

Thanks for all your help!

File: app/models/enterprise_relationship.rb
Comment (dacook) [reply app/models/enterprise_relationship.rb:89]: It looks like this method is not needed, so should be removed.

File: app/models/enterprise_relationship.rb
Comment (dacook) [inline app/models/enterprise_relationship.rb:113]: I may be ignorant about how this works, but shouldn't we only destroy the exchanges for current/new order cycles, and leave old ones alone?

Comment (dacook) [reply app/models/enterprise_relationship.rb:113]: Also, is `ExchangeDeleter` intended to cover this? That way the logic could be de-duplicated in a central place.

File: app/models/enterprise_relationship_permission.rb
Comment (dacook) [inline app/models/enterprise_relationship_permission.rb:13]: Also here ^

File: app/models/enterprise_relationship.rb
Comment (mkllnk) [reply app/models/enterprise_relationship.rb:113]: > shouldn't we only destroy the exchanges for current/new order cycles, and leave old ones alone?

Neal actually did this and I told him not do because people can change the date of order cycles. Or there's the possibility that someone creates an order cycle without dates, then permissions change and then dates are set. The order cycle would then have unpermitted products in there. And I'm not sure if these exchanges would all be duplicated on order cycle clone as well. That's much more likely.

The downside is that we lose the history of what products were in an old order cycle but we got that problem already and don't need to solve it here. I think it's better to honour the permissions correctly and solve the history problem separately, for example keeping a log or just soft-deleting exchanges. And maybe we will not allow to change the dates of passed order cycles at some point.

Comment (mkllnk) [reply app/models/enterprise_relationship.rb:113]: > is ExchangeDeleter intended to cover this? That way the logic could be de-duplicated in a central place.

It was used for this but then the logic became so simple that I decided that it would be better in the model. The ExchangeDeleter is now unused and needs deleting.

**Review by dacook — COMMENTED** [review]
Nice work, now it's whittled down to a relatively simple update. Just to make sure I'm following, a few more questions sorry:

**When an EnterpriseRelationship/Permission is deleted, also delete the associated exchange.**
1. I can see that one is `before_destroy` and the other is `after_destroy`. I'm guessing this is for a reason?
2. Can a code comment be added to explain why only clean up after `add_to_order_cycle` exchanges?

**When an ExchangeVariant is deleted, also delete other ExchangeVariants for the same Variant and OrderCycle**
I'm not familiar with this part of the system but would have guessed that there'd be only one ExchangeVariant per OrderCycle/Variant. I guess this is because there is both an incoming and outgoing ExchangeVariant.
What if the outgoing is deleted, should the corresponding incoming be deleted too?

**Comment by macanudo527** [comment]
> Nice work, now it's whittled down to a relatively simple update. Just to make sure I'm following, a few more questions sorry:
> 
> **When an EnterpriseRelationship/Permission is deleted, also delete the associated exchange.**
> 
> 1. I can see that one is `before_destroy` and the other is `after_destroy`. I'm guessing this is for a reason?

I think these should actually all be consistently `before_destroy`. I'm not sure why they are different. I think I might have been troubleshooting something.

> 2. Can a code comment be added to explain why only clean up after `add_to_order_cycle` exchanges?

Should I add this to the `spec`?

> 
> **When an ExchangeVariant is deleted, also delete other ExchangeVariants for the same Variant and OrderCycle** I'm not familiar with this part of the system but would have guessed that there'd be only one ExchangeVariant per OrderCycle/Variant. I guess this is because there is both an incoming and outgoing ExchangeVariant. What if the outgoing is deleted, should the corresponding incoming be deleted too?

I'm not sure actually. The outgoing exchange is used to apply `exchange_fees` to the processing of selling the products at the hub. The way I was thinking of it is this is the exchange between the back of the store (where deliveries are taken) to the front of the store (where the product is sold). 

It could be useful to delete the accompanying incoming exchange but this may cause unexpected behavior in the orderCycle wizard. Users first add the incoming exchanges, then the outgoing exchanges (what is offered at that hub). If they delete the outgoing exchange the incoming exchange will also be deleted. However, if they added the incoming exchange, but didn't add an outgoing exchange, the incoming exchange would still remain. The incoming exchange would only be deleted if they (added then) deleted the outgoing exchange. That seems inconsistent to me.

I could think of a case where someone might add all possible products to an order cycle and then add outgoing exchanges, save them once, delete them because they don't want to offer them, then changing their mind and going back and trying to add them, but realize they need to re-add incoming exchanges for the variants that were previously added. But, maybe that is too fringe?

Logically if you think about it, it is okay to have something come in, and not go out for whatever reason, but you can't have something go out without it ever coming in.

**Comment by dacook** [comment]
Thanks for your responses and discussion, it sounds logical to me. 

I notice that we're missing specs, did these get lost in the squash? (Sorry if it caused extra challenges like this!)
Are you able to add specs to cover these changes? And yes, an explanation of the `add_to_order_cycle` condition could go there if it makes sense.

Once there are specs, it may be clearer and easier to test if each case should be `before_` or `after_` `destroy`.

**Review by dacook — COMMENTED** [review]
Regarding  `before_` and `after_ ` `destroy`: thinking about it again I'm not sure how much it would matter. If any destroys raise an error, I assume it would abort any changes made (they would all be in the one DB transaction).

So I did a quick search of the codebase. I can see `before_destroy` used sometimes, usually a validation-type check but sometimes also to clean up or update associated records as needed. `after_destroy` is used one for either cause too, but less frequently.🤷  So I guess `before_destroy` is a better choice.. 

Also, it looks as though the specs are an older version that don't have updates made from Maikel's earlier comments, is there anything to add there?

**Comment by mkllnk** [comment]
> What if the outgoing is deleted, should the corresponding incoming be deleted too?

No. Order cycles can have multiple distributors. You can add apples and pears from several suppliers and then one distributor can have only apples in the shop and the other can have only pears in the shop. But they can also have everything in the shop. So when one shop takes a variant out, we don't want it to become unavailable for other shops. The incoming variants are like an inventory for this order cycle and then the outgoing variants define what appears in the individual shop fronts.

**Comment by macanudo527** [comment]
> Also, it looks as though the specs are an older version that don't have updates made from Maikel's earlier comments, is there anything to add there?

Are you referring to adding the "-=" to show the permission is deleted? We couldn't do that since it would involve writing an extra method that is only used in the Spec. I think the Spec should be up to date.

**Comment by macanudo527** [comment]
> So I did a quick search of the codebase. I can see `before_destroy` used sometimes, usually a validation-type check but sometimes also to clean up or update associated records as needed. `after_destroy` is used one for either cause too, but less frequently.🤷 So I guess `before_destroy` is a better choice..

It looks like `before_destroy` in `exchange_variant.rb` causes an infinite loop. Oops. I'll keep that one `after_destroy`

**Comment by macanudo527** [comment]
I think this is good to go unless there is something else you think needs to be added.

File: spec/models/enterprise_relationship_spec.rb
Comment (mkllnk) [inline spec/models/enterprise_relationship_spec.rb:320]: I think that the `*internal_only*` variables could be named `other...` to tell us that it's unrelated.

File: spec/models/enterprise_relationship_spec.rb
Comment (mkllnk) [inline spec/models/enterprise_relationship_spec.rb:313]: I personally would do the destroy within the test. And I would probably include it within an expectation:
```rb
expect {
  er.destroy
}.to change {
  ExchangeVariant.count
}.by(-2)
```

**Review by mkllnk — APPROVED** [review]
Good work. I find it very easy to read now.

**Comment by filipefurtad0** [comment]
Hey @macanudo527 ,

Thanks for this contribution. If I understand correctly, the steps to test would be:

- create a supplier enterprise with products of a given category
- grant permissions to a shop; add those products to the OC within that shop
- see that the product categories appear in the shopfront (pic 1 - shopfront, left; supplier permissions, right)
- remove the set enterprise permissions
- confirm that the product category from the supplier enterprise does not appear in the `shop for` section (pic 2 - shopfront, left; supplier permissions, right)

Pic 1

image


Pic 2

image

I believe this is it, and the test demonstrates it. I'll move to ready to go for now in case I can think of something else to test, or if you feel there is any other aspect to be verified manually.

Thanks!

**Pull request merged**
